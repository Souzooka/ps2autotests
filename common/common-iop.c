#include <stdio.h>
#include <stdarg.h>
#include <xprintf.h>

#define SCHEDF_BUFFER_SIZE 65536

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsnprintf(schedfBuffer + schedfBufferPos, SCHEDF_BUFFER_SIZE, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
