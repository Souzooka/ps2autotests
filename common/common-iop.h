#ifndef PS2AUTOTESTS_COMMON_IOP_H
#define PS2AUTOTESTS_COMMON_IOP_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _IOP
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

void schedf(const char *format, ...);
void flushschedf();

#endif
