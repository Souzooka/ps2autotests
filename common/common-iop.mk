# _____     ___ ____     ___ ____
#  ____|   |    ____|   |        | |____|
# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
#-----------------------------------------------------------------------
# Copyright 2001-2004, ps2dev - http://www.ps2dev.org
# Licenced under Academic Free License version 2.0
# Review ps2sdk README & LICENSE files for further details.
#
# $Id: Makefile.eeglobal_cpp_sample 1525 2009-01-23 00:24:13Z radad $

# Define pefixes for the various toolchain binaries.
include $(COMMON_DIR)/common-defines.mk

ASFLAGS_TARGET = -mcpu=r3000

CFLAGS_TARGET  = -miop
ASFLAGS_TARGET = -march=r3000
LDFLAGS_TARGET = -miop

# include dir
IOP_INCS := -I$(PS2SDK)/iop/include -I$(PS2SDK)/common/include -I$(COMMON_DIR) \
	-I. $(IOP_INCS)
# C compiler flags
IOP_CFLAGS := $(CFLAGS_TARGET) -O2 -G0 $(IOP_INCS) $(IOP_CFLAGS)

# linker flags
IOP_LDFLAGS := $(LDFLAGS_TARGET) -nostdlib -L$(PS2SDK)/iop/lib $(IOP_LDFLAGS)

# asssembler flags
IOP_ASFLAGS := $(ASFLAGS_TARGET) -EL -G0 $(IOP_ASFLAGS)

# link with following libraries (libs need to be defined multiple times in order for linking to work!!)
IOP_LIBS += -lkernel -lgcc

# Externally defined variables: IOP_BIN, IOP_OBJS, IOP_LIB

%.o : %.c
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.S
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.s
	$(IOP_AS) $(IOP_ASFLAGS) $< -o $@

%.o : %.lst
	echo "#include \"irx_imports.h\"" > build-imports.c
	cat $< >> build-imports.c
	$(IOP_CC) $(IOP_CFLAGS) -c build-imports.c -o $@
	rm -f build-imports.c

%.irx : %.o $(EXTRA_OBJS)
	$(IOP_CC) $(IOP_LDFLAGS) -o $@ $< $(EXTRA_OBJS) $(IOP_LIBS)

all: $(TARGETS:=.irx)

clean:
	-$(RM) -f $(TARGETS:=.irx) $(TARGETS:=.o)

rebuild: clean all
