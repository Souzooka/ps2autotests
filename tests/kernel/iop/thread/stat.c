#include <common-iop.h>
#include <stdio.h>
#include <tamtypes.h>
#include <thbase.h>
#include <thsemap.h>
#include <thevent.h>
#include <thmsgbx.h>
#include "thread-common.h"

#define STACK_SIZE 0x800

void nullThreadProc(u32 param) {
	//Do nothing
}

void sleepThreadProc(u32 param) {
	SleepThread();
}

void delayThreadProc(u32 param) {
	DelayThread(~0);
}

void semaThreadProc(u32 param) {
	iop_sema_t dummySemaInfo = {};
	dummySemaInfo.attr    = 0;
	dummySemaInfo.option  = 0;
	dummySemaInfo.initial = 0;
	dummySemaInfo.max     = 1;
	u32 dummySema = CreateSema(&dummySemaInfo);
	WaitSema(dummySema);
}

void eventFlagThreadProc(u32 param) {
	iop_event_t dummyEventFlagInfo = {};
	dummyEventFlagInfo.attr = 0;
	dummyEventFlagInfo.bits = 0;
	u32 dummyEventFlag = CreateEventFlag(&dummyEventFlagInfo);
	WaitEventFlag(dummyEventFlag, ~0, WEF_CLEAR | WEF_AND, NULL);
}

void messageBoxThreadProc(u32 param) {
	iop_mbx_t dummyMessageBoxInfo = {};
	dummyMessageBoxInfo.attr = 0;
	dummyMessageBoxInfo.option = 0;
	u32 dummyMessageBox = CreateMbx(&dummyMessageBoxInfo);
	ReceiveMbx(NULL, dummyMessageBox);
}

void printThreadStatus(int threadId) {
	iop_thread_info_t threadStat;
	memset(&threadStat, 0, sizeof(iop_thread_info_t));
	int result = ReferThreadStatus(threadId, &threadStat);
	if (result >= 0) {
		schedf("succeeded -> result: %d, init prio: %02x current prio: %02x, status: %02x", 
			result, threadStat.initPriority, threadStat.currentPriority, threadStat.status);
		if (threadStat.status & THS_WAIT) {
			schedf(", wait type: %d", threadStat.waitType);
		}
		schedf("\n");
	} else {
		schedf("failed -> result: %d\n", result);
	}
}

void testWaitingState(const char *name, void *threadProc, u32 threadPrio) {
	int threadId = createTestThread(threadProc, threadPrio, STACK_SIZE);

	StartThread(threadId, 0);
	schedf("  stat after start (%s): ", name);
	printThreadStatus(threadId);
	
	TerminateThread(threadId);
	DeleteThread(threadId);
}

int _start(int argc, char **argv) {
	printf("-- TEST BEGIN\n");
	
	int currentThreadPriority = getThreadPriority(TH_SELF);
	
	{
		schedf("self thread:\n");
		
		schedf("  stat (with TH_SELF): ");
		printThreadStatus(TH_SELF);
		
		schedf("  stat (with current tid): ");
		printThreadStatus(GetThreadId());
		
		flushschedf();
	}
	
	{
		schedf("low prio thread:\n");
		
		int threadId = createTestThread((void*)&nullThreadProc, currentThreadPriority + 1, STACK_SIZE);
		schedf("  stat before start: ");
		printThreadStatus(threadId);

		StartThread(threadId, 0);
		schedf("  stat after start: ");
		printThreadStatus(threadId);
		
		//Let the thread execute (terminate doesn't seem to work properly otherwise)
		DelayThread(1);
		
		TerminateThread(threadId);
		schedf("  stat after terminate: ");
		printThreadStatus(threadId);
		
		DeleteThread(threadId);

		flushschedf();
	}

	{
		schedf("waiting thread:\n");
		
		testWaitingState("sleep", (void*)&sleepThreadProc, currentThreadPriority - 1);
		testWaitingState("delay", (void*)&delayThreadProc, currentThreadPriority - 1);
		testWaitingState("semaphore", (void*)&semaThreadProc, currentThreadPriority - 1);
		testWaitingState("event flag", (void*)&eventFlagThreadProc, currentThreadPriority - 1);
		testWaitingState("message box", (void*)&messageBoxThreadProc, currentThreadPriority - 1);
		
		flushschedf();
	}
	
	{
		schedf("corner cases:\n");
		
		schedf("  stat with invalid thread id: ");
		printThreadStatus(~0);
		
		flushschedf();
	}
	
	printf("-- TEST END\n");
	
	return 0;
}
