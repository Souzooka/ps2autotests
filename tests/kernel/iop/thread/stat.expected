-- TEST BEGIN
self thread:
  stat (with TH_SELF): succeeded -> result: 0, init prio: 08 current prio: 08, status: 01
  stat (with current tid): succeeded -> result: 0, init prio: 08 current prio: 08, status: 01
low prio thread:
  stat before start: succeeded -> result: 0, init prio: 09 current prio: 00, status: 10
  stat after start: succeeded -> result: 0, init prio: 09 current prio: 09, status: 02
  stat after terminate: succeeded -> result: 0, init prio: 09 current prio: 09, status: 10
waiting thread:
  stat after start (sleep): succeeded -> result: 0, init prio: 07 current prio: 07, status: 04, wait type: 1
  stat after start (delay): succeeded -> result: 0, init prio: 07 current prio: 07, status: 04, wait type: 2
  stat after start (semaphore): succeeded -> result: 0, init prio: 07 current prio: 07, status: 04, wait type: 3
  stat after start (event flag): succeeded -> result: 0, init prio: 07 current prio: 07, status: 04, wait type: 4
  stat after start (message box): succeeded -> result: 0, init prio: 07 current prio: 07, status: 04, wait type: 5
corner cases:
  stat with invalid thread id: failed -> result: -407
-- TEST END
